src/clients/c++/perf_client/load_manager.cc

```diff
@@ -213,33 +306,297 @@ LoadManager::InitManagerInputs(
-        local_manager->input_string_buf_[i] = string_data;
+        input_string_buf_[i] = string_data;
       }
     } else {
       for (size_t i = 0; i < max_batch1_num_strings; i++) {
-        local_manager->input_string_buf_[i] = GetRandomString(string_length);
+        input_string_buf_[i] = GetRandomString(string_length);
       }
     }
   }
 
   // Reserve the required vector space
-  local_manager->threads_stat_.reserve(local_manager->max_threads_);
+  threads_stat_.reserve(max_threads_);
+
+  return nic::Error::Success;
+}
+
+nic::Error
+LoadManager::InitSharedMemory()
+{
+  nic::Error err;
+
+  RETURN_IF_ERROR(
+      factory_->CreateSharedMemoryControlContext(&shared_memory_ctx_));
+  std::unique_ptr<nic::InferContext> ctx;
+  RETURN_IF_ERROR(factory_->CreateInferContext(&ctx));
 
-  *manager = std::move(local_manager);
+  // Allocate the shared memory for outputs
+  for (const auto& output : ctx->Outputs()) {
+    int64_t batch1_bytesize = ctx->ByteSize(output->Dims(), output->DType());
+    if (batch1_bytesize < 0) {
+      batch1_bytesize = output_shm_size_;
+    }
+    uint8_t* output_shm_ptr;
+    size_t alloc_size = batch1_bytesize * batch_size_;
+    if (shared_memory_type_ == SharedMemoryType::SYSTEM_SHARED_MEMORY) {
+      std::string shm_key("/" + output->Name());
+      int shm_fd_op;
+      RETURN_IF_ERROR(
+          CreateSharedMemoryRegion(shm_key, alloc_size, &shm_fd_op));
+      RETURN_IF_ERROR(
+          MapSharedMemory(shm_fd_op, 0, alloc_size, (void**)&output_shm_ptr));
+
+      io_shm_size_[output->Name()] = alloc_size;
+      shared_memory_regions_[output->Name()] = output_shm_ptr;
+
+      RETURN_IF_ERROR(shared_memory_ctx_->RegisterSharedMemory(
+          output->Name(), shm_key, 0, alloc_size));
+    } else {
+#if TRTIS_ENABLE_GPU
+      cudaError_t cuda_err = cudaMalloc((void**)&output_shm_ptr, alloc_size);
+      if (cuda_err != cudaSuccess) {
+        return nic::Error(
+            ni::RequestStatusCode::INTERNAL,
+            "unable to allocate memory of " + std::to_string(alloc_size) +
+                "bytes on gpu for output " + output->Name());
+      }
+      io_shm_size_[output->Name()] = alloc_size;
+      shared_memory_regions_[output->Name()] = output_shm_ptr;
+
+      cudaIpcMemHandle_t cuda_handle;
+      RETURN_IF_ERROR(CreateCUDAIPCHandle(&cuda_handle, (void*)output_shm_ptr));
+      // Using GPU with device id 0
+      RETURN_IF_ERROR(shared_memory_ctx_->RegisterCudaSharedMemory(
+          output->Name(), cuda_handle, alloc_size, 0));
+#endif  // TRTIS_ENABLE_GPU
+    }
+  }
+
+  for (const auto& input : ctx->Inputs()) {
+    if (input->Shape().empty()) {
+      auto it = input_shapes_.find(input->Name());
+      if (it != input_shapes_.end()) {
+        input->SetShape(it->second);
+      }
+    }
+    std::vector<char> serialized_data;
+    const uint8_t* data;
+    size_t batch1_bytesize;
+    if (input->DType() != ni::DataType::TYPE_STRING) {
```

Comment:
What about when using variable dims inputs?

Reply:
We have already used SetShape() with the user given shapes, hence will receive a correct ByteSize()